File: LeetCode/207. 课程表.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        hasCircle = False
        visited = [False for i in range(numCourses)]
        onPath = [False for i in range(numCourses)]
        graph = [[] for i in range(numCourses)]
        for (cur, pre) in prerequisites:
            graph[pre].append(cur)
        def traverse(start):
            nonlocal hasCircle,visited,onPath,graph
            if onPath[start]:
                hasCircle = True
            if hasCircle or visited[start]:
                return
            onPath[start] = True
            visited[start] = True
            for x in graph[start]:
                traverse(x)
            onPath[start] = False
        for i in range(numCourses):
            traverse(i)
        return not hasCircle

File: LeetCode/207. 课程表.rs
impl Solution {
    pub fn can_finish(num_courses: i32, prerequisites: Vec<Vec<i32>>) -> bool {
        let num_courses = num_courses as usize;
        let mut has_circle = Box::new(false);
        let mut visited = vec![false; num_courses];
        let mut on_path = vec![false; num_courses];
        let mut graph = vec![Vec::new(); num_courses];
        for pair in &prerequisites[..] {
            if let &[cur, pre] = &pair[..] {
                graph[pre as usize].push(cur);
            }
        }
        fn traverse(mut visited: &mut Vec<bool>, mut on_path: &mut Vec<bool>, graph: &Vec<Vec<i32>>, mut has_circle: &mut Box<bool>, start:i32) {
            let idx = start as usize;
            if on_path[idx] {
                **has_circle = true;
            }
            if **has_circle || visited[idx] {
                return;
            }
            on_path[idx] = true;
            visited[idx] = true;
            for &x in &graph[idx] {
                traverse(&mut visited, &mut on_path, &graph, &mut has_circle, x);
            }
            on_path[idx] = false;
        }
        for x in 0..num_courses {
            traverse(&mut visited, &mut on_path, &graph, &mut has_circle, x as i32);
        }
        return !*has_circle;
    }
}
